pkg/encoding/kzgEncoder/multiframe.go

```diff
@@ -0,0 +1,212 @@
+// we can also pseudo randomness generated locally, but we have to ensure no adversary can manipulate it
+// Hashing everything takes about 1ms, so Fiat Shamir transform does not incur much cost
+func GenRandomFactor(samples []Sample) (bls.Fr, error) {
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+
+	for _, sample := range samples {
+		err := enc.Encode(sample.Commitment)
+		if err != nil {
+			return bls.ZERO, err
+		}
+	}
+
+	var randomFr bls.Fr
+
+	err := bls.HashToSingleField(&randomFr, buffer.Bytes())
+	if err != nil {
+		return bls.ZERO, err
+	}
+
+	return randomFr, nil
+}
+
+// Every sample has its own randomness, even though multiple samples can come from identical blob
+// Randomnesss for each sample is computed by repeatedly raising the power of the root randomness
+func GenRandomnessVector(samples []Sample) ([]bls.Fr, error) {
+	// root randomness
+	r, err := GenRandomFactor(samples)
+	if err != nil {
+		return nil, err
+	}
+
+	n := len(samples)
+
+	randomsFr := make([]bls.Fr, n)
+	bls.CopyFr(&randomsFr[0], &r)
+
+	// power of r
+	for j := 0; j < n-1; j++ {
+		bls.MulModFr(&randomsFr[j+1], &randomsFr[j], &r)
+	}
+	return randomsFr, nil
+}
+
+// the rhsG1 comprises of three terms, see https://ethresear.ch/t/a-universal-verification-equation-for-data-availability-sampling/13240/1
+func genRhsG1(samples []Sample, randomsFr []bls.Fr, m int, params rs.EncodingParams, ks *kzg.KZGSettings, proofs []bls.G1Point) (*bls.G1Point, error) {
+	n := len(samples)
+	commits := make([]bls.G1Point, m)
+	D := params.ChunkLen
+
+	var tmp bls.Fr
+
+	// first term
+	// get coeffs to compute the aggregated commitment
+	// note the coeff is affected by how many chunks are validated per blob
+	// if x chunks are sampled from one blob, we need to compute the sum of all x random field element corresponding to each sample
+	aggCommitCoeffs := make([]bls.Fr, m)
+	for k := 0; k < n; k++ {
+		s := samples[k]
+		row := s.RowIndex
+		bls.AddModFr(&aggCommitCoeffs[row], &aggCommitCoeffs[row], &randomsFr[k])
+		bls.CopyG1(&commits[row], &s.Commitment)
+	}
+
+	aggCommit := bls.LinCombG1(commits, aggCommitCoeffs)
+
+	// second term
+	// compute the aggregated interpolation polynomial
+	aggPolyCoeffs := make([]bls.Fr, D)
+
+	// we sum over the weighted coefficients (by the random field element) over all D monomial in all n samples
+	for k := 0; k < n; k++ {
+		coeffs := samples[k].Coeffs
+
+		rk := randomsFr[k]
+		// for each monomial in a given polynomial, multiply its coefficient with the corresponding random field,
+		// then sum it with others. Given ChunkLen (D) is identical for all samples in a subBatch.
+		// The operation is always valid.
+		for j := uint64(0); j < D; j++ {
+			bls.MulModFr(&tmp, &coeffs[j], &rk)
+			bls.AddModFr(&aggPolyCoeffs[j], &aggPolyCoeffs[j], &tmp)
+		}
+	}
+
+	// All samples in a subBatch has identical chunkLen
+	aggPolyG1 := bls.LinCombG1(ks.Srs.G1[:D], aggPolyCoeffs)
+
+	// third term
```

Comment:
btw I checked the code of these 3 terms against the research post and it looks good (quite impressive you nail those details rigorously!), except the 3rd term which I cannot follow (need to understand the coset shifting stuff a bit more)